File: DS_SPECIFIC_CODE/LinkedList/C++/SinglyCL.cpp
//  SINGLY CIRCULAR LINKED LIST

#include<iostream>

using namespace std;

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node* PNODE;

class SinglyCL
{
    private:
        PNODE first;
        PNODE last ;
        int iCount;

    public:
        SinglyCL()
        {
            cout<<"Object of SinglyCL gets created\n";
            this-> first = NULL;
            this-> last = NULL;
            this -> iCount = 0;
        }

        void InsertFirst(int No)
        {
            PNODE newn = NULL;
            newn = new NODE;

            newn -> data = No;
            newn -> next = NULL;

            if(this->first ==NULL && this->last == NULL)
            {
                this->first= newn;
                this->last = newn;
                this->last->next = this->first;
            }
            else
            {
                newn->next = this->first;
                this->first = newn;
                this->last->next = this->first;

            }

            this->iCount++;
        }

        void InsertLast(int No)
        {
            PNODE newn = NULL;
            newn = new NODE;

            newn -> data = No;
            newn -> next = NULL;

            if(this->first == NULL && this->last == NULL)
            {
                this->first = newn;
                this->last = newn;
            }
            
            else
            {
                this->last->next = newn;
                this->last = newn;
                this->last -> next = this->first;
            }

            iCount++;

        }

        void DeleteFirst()
        {
            if(this->first == NULL && this->last ==NULL)
            {
                return;
            }

            else if(this->first == this->last)
            {
                delete this -> first;
                delete this -> last;
                this->first = NULL;
                this->last = NULL;
            }

            else
            {
                this->first = this ->first -> next;
                delete this-> last -> next;
                this -> last -> next = this -> first;
            }

            iCount --;
        }

        void DeleteLast()
        {
            PNODE temp = NULL;

            if(this->first == NULL && this->last ==NULL)
            {
                return;
            }

            else if(this->first == this->last)
            {
                delete this -> first;
                delete this -> last;
                this->first = NULL;
                this->last = NULL;
            }
            
            else
            {
                temp = this->first;

                while(temp->next != this -> last)
                {
                    temp = temp -> next;
                }

                delete this -> last;
                this->last =  temp;
            }

            this->last->next = this -> first;

            iCount --;
        }

        void Display()
        {
            PNODE temp = NULL;

            temp = this->first;

            do
            {
                cout<<"| "<<temp->data<<" | ->";
                temp= temp->next;
            } while(temp != this->last->next);
            cout<<"\n";
        }

        int Count()
        {
            return iCount;
        }

        void InsertAtPos(int No, int pos)
        {
            PNODE temp = NULL;
            PNODE newn = NULL;
            
            int iCnt = 0;

            if(pos<1 || pos>iCount+1)
            {
                cout<<"Invalid position";
                return;
            }

            if(pos == 1)
            {
                this->InsertFirst(No);
            }

            else if(pos == iCount+1)
            {
                this->InsertLast(No);
            }

            else
            {
                temp = this->first;

                newn = new NODE;

                newn -> data = No;
                newn -> next = NULL;

                for(iCnt = 1; iCnt<pos-1; iCnt++)
                {
                    temp = temp -> next;
                }

                newn -> next = temp -> next;
                temp -> next = newn;

                iCount++;
            }
        }


        void DeleteAtPos(int pos)
        {
            PNODE temp = NULL;
            PNODE target = NULL;
            
            int iCnt = 0;

            if(pos<1 || pos>iCount)
            {
                cout<<"Invalid position";
                return;
            }

            if(pos == 1)
            {
                this->DeleteFirst();
            }

            else if(pos == iCount)
            {
                this->DeleteLast();
            }

            else
            {
                temp = this->first;

                for(iCnt = 1; iCnt<pos-1; iCnt++)
                {
                    temp = temp -> next;
                }

                target = temp -> next;
                temp -> next = target -> next;
                delete target;

                iCount --;  
            }
        }

};

int main()
{

    SinglyCL obj;
    int iRet  = 0;

    obj.InsertFirst(51);
    obj.InsertFirst(21);
    obj.InsertFirst(11);

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are: "<<iRet<<"\n";

    obj.InsertLast(101);
    obj.InsertLast(111);
    obj.InsertLast(121);

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are: "<<iRet<<"\n";

    obj.DeleteFirst();

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are: "<<iRet<<"\n";

    obj.DeleteLast();

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are: "<<iRet<<"\n";

    obj.InsertAtPos(105,4);

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are: "<<iRet<<"\n";
    
    obj.DeleteAtPos(4);

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of nodes are: "<<iRet<<"\n";

    return 0;
}

File: DS_SPECIFIC_CODE/LinkedList/Java/SinglyCL.java
// SINGLY CIRCULAR LIKEDLIST

class node
{
    public int data;
    public node next;

    public node(int No)
    {
        this.data = No;
        this.next = null;
    }

}

class SinglyCL
{
    private node first;
    private node last;
    private int iCount;

    public SinglyCL()
    {

        this.first = null;
        this.last = null;
        this.iCount = 0;
    }

    public void InsertFirst(int No)
    {
        node newn = null;
        newn = new node(No);

        if(this.first == null && this.last == null)
        {
            this.first = newn;
            this.last = newn;
        }

        else
        {
            newn.next = this.first;
            this.first = newn;
            this.last.next = this.first;
        }
        
        this.iCount++;
    }

    public void InsertLast(int No)
    {
        node newn = null;

        newn = new node(No);

        if(this.first == null && this.last == null)
        {
            this.first = newn;
            this.last = newn;
        }

        else
        {
            this.last.next = newn;
            this.last = newn;
        }

        this.last.next = this.first;
        this.iCount++;
    }

    public void DeleteFirst()
    {
        if(this.first == null && this.last==null)
        {
            return;
        }
        else if(this.first == this.last)
        {
            this.first = null;
            this.last = null;
        }

        else
        {
            this.first = this.first.next;
            this.last.next = this.first;
        }

        System.gc();

        this.iCount--;
    }

    public void DeleteLast()
    {
        node temp = null;
        if(this.first == null && this.last == null)
        {
            return;
        }

        else if(this.first == this.last)
        {
            this.first = null;
            this.last = null;
        }

        else
        {
            temp = this.first;

            while(temp.next != this.last)
            {
                temp = temp.next;
            }

            this.last = null;
            this.last = temp;
        }

        this.last.next = this.first;

        System.gc();

        this.iCount--;

    }

    public void Display()
    {
        node temp = null;

        temp = this.first;

        do
        {
            System.out.print("| "+temp.data+"| ->");
            temp = temp.next;
        }while(temp != this.last.next);

        System.out.println();
    }

    public int Count()
    {
        return this.iCount;
    }

    public void InsertAtPos(int No, int Pos)
    {
        int iCnt = 0;
        node newn = null;
        node temp = null;

        if(Pos<1 || Pos>this.iCount+1)
        {
            System.out.println("Invalid Position");
            return;
        }

        if(Pos == 1)
        {
            this.InsertFirst(No);
        }

        else if(Pos == iCount+1)
        {
            this.InsertLast(No);
        }

        else
        {
           newn = new node(No);
           
           temp = this.first;

           for(iCnt = 1; iCnt<Pos-1; iCnt++)
           {
                temp = temp.next;
           }

           newn.next= temp.next;
           temp.next = newn;

           this.iCount++;
        }
    }

    public void DeleteAtPos(int Pos)
    {
        int iCnt = 0;
        node temp = null;
        node target = null;

        if(Pos<1 || Pos>this.iCount)
        {
            System.out.println("Invalid Position");
            return;
        }

        if(Pos == 1)
        {
            this.DeleteFirst();
        }

        else if(Pos == iCount)
        {
            this.DeleteLast();
        }

        else
        {

            temp = this.first;

           for(iCnt = 1; iCnt<Pos-1; iCnt++)
           {
                temp = temp.next;
           }

           target = temp.next;
           temp.next = target.next;

           target = null;
           System.gc();
           this.iCount--;
        }
    }

}

class program447
{
    public static void main(String A [])
    {
        SinglyCL obj = null;
        int iRet = 0;

        obj = new SinglyCL();

        obj.InsertFirst(51);
        obj.InsertFirst(21);
        obj.InsertFirst(11);

        obj.Display();
        
        iRet = obj.Count();
        System.out.println("The Number of nodes are :"+iRet);

        obj.InsertLast(101);
        obj.InsertLast(111);
        obj.InsertLast(121);

        obj.Display();
        
        iRet = obj.Count();
        System.out.println("The Number of nodes are :"+iRet);

        obj.DeleteFirst();

        obj.Display();
        
        iRet = obj.Count();
        System.out.println("The Number of nodes are :"+iRet);

        obj.DeleteLast();

        obj.Display();
        
        iRet = obj.Count();
        System.out.println("The Number of nodes are :"+iRet);

        obj.InsertAtPos(105, 4);

        obj.Display();
        
        iRet = obj.Count();
        System.out.println("The Number of nodes are :"+iRet);

        obj.DeleteAtPos(4);

        obj.Display();
        
        iRet = obj.Count();
        System.out.println("The Number of nodes are :"+iRet);
        
        // Important for memory deallocation
        obj=null;
        System.gc();

    }
}